### PR TITLE
formatter: add custom date formatting to downstream cert start and end dates

### PR DESCRIPTION
**Pull request by esmet**
Commit Message: formatter: add custom date formatting to downstream cert start and end dates
Additional Description: refactor and reuse existing behavior from START_TIME
Risk Level: low (new behavior, though existing code is refactored)
Testing: refactored existing unit test, added test cases for existing behaviors to new fields
Docs Changes: Updated docs/root/configuration/observability/access_log/usage.rst to mention formatting, refers back to START_TIME
Release Notes: formatter: added support for custom date formatting to %DOWNSTREAM_PEER_CERT_V_START% and %DOWNSTREAM_PEER_CERT_V_END%, similar to %START_TIME%.
Fixes: #14501

### CODE REVIEW

File: source/common/formatter/substitution_formatter.cc [hunk 1]
```diff
@@ -378,10 +378,38 @@ std::vector<FormatterProviderPtr> SubstitutionFormatParser::parse(const std::str
                                      : "";
         // Validate the input specifier here. The formatted string may be destined for a header, and
         // should not contain invalid characters {NUL, LR, CF}.
-        if (std::regex_search(args, getStartTimeNewlinePattern())) {
+        if (std::regex_search(args, getDateFormatNewlinePattern())) {
           throw EnvoyException("Invalid header configuration. Format string contains newline.");
         }
         formatters.emplace_back(FormatterProviderPtr{new StartTimeFormatter(args)});
+      } else if (absl::StartsWith(token, "DOWNSTREAM_PEER_CERT_V_START")) {
```
Comment (mattklein123) [inline source/common/formatter/substitution_formatter.cc:385]: It seems like all of this code is going to be identical for START_TIME, here, etc. Please factor all of this and the formatters below into a single implementation for when you push the full PR with docs, etc.

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks for working on this!

/wait

File: source/common/formatter/substitution_formatter.cc (hunk 1 above)
Comment (esmet) [reply source/common/formatter/substitution_formatter.cc:378]: I went ahead and de-duped this logic by converting the StartTimeFormatter to SystemTimeFormatter, and keeping several subclass implementations that get used (and unit tested) independently.

File: docs/root/configuration/observability/access_log/usage.rst [hunk 2]
```diff
@@ -541,12 +541,18 @@ The following command operators are supported:
   TCP
     The validity start date of the client certificate used to establish the downstream TLS connection.
 
+  DOWNSTREAM_PEER_CERT_V_START can be customized using a `format string <https://en.cppreference.com/w/cpp/io/manip/put_time>`_.
+  See START_TIME for additional format specifiers and examples.
```
Comment (esmet) [inline docs/root/configuration/observability/access_log/usage.rst:545]: Let me know if this level of documentation is clear/acceptable.

File: source/common/formatter/substitution_formatter.cc [hunk 3]
```diff
@@ -43,8 +43,8 @@ void truncate(std::string& str, absl::optional<uint32_t> max_length) {
   str = str.substr(0, max_length.value());
 }
 
-// Matches newline pattern in a StartTimeFormatter format string.
-const std::regex& getStartTimeNewlinePattern() {
+// Matches newline pattern in a system time format string (e.g. start time)
+const std::regex& getDateFormatNewlinePattern() {
```
Comment (esmet) [inline source/common/formatter/substitution_formatter.cc:47]: For consistency I should probably rename this to `getSystemTimeNewlinePattern`

File: docs/root/version_history/current.rst [hunk 4]
```diff
@@ -67,6 +67,7 @@ New Features
 * config: added ability to flush stats when the admin's :ref:`/stats endpoint <operations_admin_interface_stats>` is hit instead of on a timer via :ref:`stats_flush_on_admin <envoy_v3_api_field_config.bootstrap.v3.Bootstrap.stats_flush_on_admin>`.
 * config: added new runtime feature `envoy.features.enable_all_deprecated_features` that allows the use of all deprecated features.
 * formatter: added new :ref:`text_format_source <envoy_v3_api_field_config.core.v3.SubstitutionFormatString.text_format_source>` field to support format strings both inline and from a file.
+* formatter: added support for custom date formatting to %DOWNSTREAM_PEER_CERT_V_START% and %DOWNSTREAM_PEER_CERT_V_END%, similar to %START_TIME%.
```
Comment (mattklein123) [inline docs/root/version_history/current.rst:70]: Please ref link to the relevant fields.

File: source/common/formatter/substitution_formatter.cc [hunk 5]
```diff
@@ -370,18 +370,25 @@ std::vector<FormatterProviderPtr> SubstitutionFormatParser::parse(const std::str
...
-        if (std::regex_search(args, getStartTimeNewlinePattern())) {
-          throw EnvoyException("Invalid header configuration. Format string contains newline.");
-        }
         formatters.emplace_back(FormatterProviderPtr{new StartTimeFormatter(args)});
+      } else if (absl::StartsWith(token, "DOWNSTREAM_PEER_CERT_V_START")) {
+        const size_t parameters_length = token.length() - (DownstreamPeerCertVStartParamStart + 1);
+        const std::string args =
+            token[DownstreamPeerCertVStartParamStart - 1] == '('
+                ? token.substr(DownstreamPeerCertVStartParamStart, parameters_length)
+                : "";
```
Comment (mattklein123) [inline source/common/formatter/substitution_formatter.cc:383]: Can this part be somehow pushed down into the shared parser/formatter? It's pretty confusing and could use more comments anyway. Until this code is rewritten to use an actual parser grammar I would err on the side of more comments.

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks LGTM with small comments.

/wait

File: source/common/formatter/substitution_formatter.cc (hunk 5 above)
Comment (esmet) [reply source/common/formatter/substitution_formatter.cc:383]: Sounds right. I'll add a shared format parsing function and push the logic into each constructor, which should allow this to become an easy one liner.

**Comment by mattklein123** [comment]
Thanks this looks good. Can you fix the docs build?

/wait

**Comment by esmet** [comment]
I found and fixed the docs issue locally so CI should be happy this time around.

**Review by mattklein123 — APPROVED** [review]
Thanks!

**Pull request merged**